Mathlib/CategoryTheory/Sites/Coherent/RegularSheaves.lean

```diff
@@ -26,65 +27,262 @@ namespace CategoryTheory
+    P.map π.op ≫ P.map (pullback.snd).op := by
+  simp only [← Functor.map_comp, ← op_comp, pullback.condition]
+
+lemma mapToEqualizer_eq_comp (P : Cᵒᵖ ⥤ Type*) {X B : C} (π : X ⟶ B)
+    [HasPullback π π] : MapToEqualizer P π pullback.fst pullback.snd pullback.condition =
+    equalizer.lift (P.map π.op) (equalizerCondition_iff_isIso_lift_w P π) ≫
+    (Types.equalizerIso _ _).hom := by
+  rw [← Iso.comp_inv_eq (α := Types.equalizerIso _ _)]
+  apply equalizer.hom_ext
+  aesop
+
+/-- An alternative phrasing of the explicit equalizer condition, using more categorical language. -/
+theorem equalizerCondition_iff_isIso_lift (P : Cᵒᵖ ⥤ Type*) : EqualizerCondition P ↔
+    ∀ (X B : C) (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π], IsIso
+    (equalizer.lift (P.map π.op) (equalizerCondition_iff_isIso_lift_w P π)) := by
+  constructor
+  · intro hP X B π _ _
+    have h := hP.bijective_mapToEqualizer_pullback _ X B π
+    rw [← isIso_iff_bijective, mapToEqualizer_eq_comp] at h
+    exact IsIso.of_isIso_comp_right (equalizer.lift (P.map π.op)
+      (equalizerCondition_iff_isIso_lift_w P π))
+      (Types.equalizerIso _ _).hom
+  · intro hP
+    apply EqualizerCondition.mk
+    intro X B π _ _
+    rw [mapToEqualizer_eq_comp, ← isIso_iff_bijective]
+    infer_instance
+
+/-- `P` satisfies the equalizer condition iff its precomposition by an equivalence does. -/
+theorem equalizerCondition_iff_of_equivalence (P : Cᵒᵖ ⥤ D)
+    (e : C ≌ E) : EqualizerCondition P ↔ EqualizerCondition (e.op.inverse ⋙ P) :=
+  ⟨fun h ↦ equalizerCondition_precomp_of_preservesPullback P e.inverse h, fun h ↦
+    equalizerCondition_of_natIso (e.op.funInvIdAssoc P)
+      (equalizerCondition_precomp_of_preservesPullback (e.op.inverse ⋙ P) e.functor h)⟩
+
+open WalkingParallelPair WalkingParallelPairHom in
+theorem parallelPair_initial_of_zigzag_zero {X Y : C} (f g : X ⟶ Y)
+    [∀ (Z : C), Nonempty (CostructuredArrow (parallelPair f.op g.op) ⟨Z⟩)]
+    (h : ∀ {Z : C} (i j : Z ⟶ Y), Zigzag
+      (⟨zero, ⟨()⟩, i.op⟩ : CostructuredArrow (parallelPair f.op g.op) ⟨Z⟩) ⟨zero, ⟨()⟩, j.op⟩) :
+    (parallelPair f.op g.op).Initial where
+  out _ := by
+    let H := parallelPair f.op g.op
+    apply zigzag_isConnected
+    rintro ⟨⟨_ | _⟩, _, ⟨i⟩⟩ ⟨⟨_ | _⟩, _, ⟨j⟩⟩
+    · exact h i j
+    · exact (h i (j ≫ (H.map right).unop)).trans (Relation.ReflTransGen.single
+        (Or.inl ⟨CostructuredArrow.homMk right rfl⟩))
+    · refine (?_ : Zigzag _ _).trans (h (i ≫ (H.map left).unop) j)
+      exact Relation.ReflTransGen.single (Or.inr ⟨CostructuredArrow.homMk left rfl⟩)
+    · refine ((?_ : Zigzag _ _).trans
+        (h (i ≫ (H.map left).unop) (j ≫ (H.map right).unop))).trans
+        (Relation.ReflTransGen.single (Or.inl ⟨CostructuredArrow.homMk right rfl⟩))
+      exact Relation.ReflTransGen.single (Or.inr ⟨CostructuredArrow.homMk left rfl⟩)
+
+open WalkingParallelPair WalkingParallelPairHom in
+theorem parallelPair_pullback_initial {X B : C} (π : X ⟶ B) (c : PullbackCone π π)
```

Comment:
For convenience, the following API for the category attached to a presieve could be introduced:
```lean
-- change the definition of `Presieve.diagram` to make this the source category
abbrev _root_.CategoryTheory.Presieve.category {X : C} (P : Presieve X) :=
  FullSubcategory fun f : Over X => P f.hom

abbrev _root_.CategoryTheory.Presieve.categoryMk {X : C} (P : Presieve X)
  {Y : C} (f : Y ⟶ X) (hf : P f) : P.category := ⟨Over.mk f, hf⟩
```

Then, the phrasing of this lemma can be slightly improved:
```lean
theorem parallelPair_pullback_initial {X B : C} (π : X ⟶ B)
    (c : PullbackCone π π) (hc : IsLimit c) :
    (parallelPair (C := (Sieve.ofArrows (fun (_ : Unit) => X) (fun _ => π)).arrows.categoryᵒᵖ)
    (Y := op ((Presieve.categoryMk _ (c.fst ≫ π) ⟨_, c.fst, π, ofArrows.mk (), rfl⟩)))
    (X := op ((Presieve.categoryMk _ π (Sieve.ofArrows_mk _ _ Unit.unit))))
    (Quiver.Hom.op (Over.homMk c.fst)) (Quiver.Hom.op (Over.homMk c.snd c.condition.symm))).Initial := by
  apply Limits.parallelPair_initial_mk
  all_goals sorry
```
(I am thinking that we could also introduce an avbreviation `Sieve.ofSingleArrow π`.)

Reply:
Ok, I've filled in this sorry. I'll clean everything up tomorrow,